src/main/resources/data/computercraft/lua/rom/programs/fun/speaker.lua

```diff
@@ -40,12 +48,60 @@ elseif cmd == "play" then
         error(err, 0)
     end
 
+    local start = handle.read(4)
+    local pcm = false
+    if start == "RIFF" then
+        handle.read(4)
+        if handle.read(8) ~= "WAVEfmt " then
+            handle.close()
+            error("Could not play audio: Unsupported WAV file", 0)
+        end
+
+        local fmtsize = ("<I4"):unpack(handle.read(4))
+        local fmt = handle.read(fmtsize)
+        local format, channels, rate, _, _, bits = ("<I2I2I4I4I2I2"):unpack(fmt)
+        if not ((format == 1 and bits == 8) or (format == 0xFFFE and bits == 1)) then
+            handle.close()
+            error("Could not play audio: Unsupported WAV file", 0)
+        end
+        if channels ~= 1 or rate ~= 48000 then
+            print("Warning: This WAV file may not play correctly.")
+        end
+        if format == 0xFFFE then
+            local guid = fmt:sub(25)
+            if guid ~= "\x3A\xC1\xFA\x38\x81\x1D\x43\x61\xA4\x0D\xCE\x53\xCA\x60\x7C\xD1" then -- DFPWM format GUID
+                handle.close()
+                error("Could not play audio: Unsupported WAV file", 0)
+            end
+        else
+            pcm = true
+        end
+
+        repeat
+            local chunk = handle.read(4)
+            if chunk == nil then
+                handle.close()
+                error("Could not play audio: Invalid WAV file", 0)
+            elseif chunk ~= "data" then -- Ignore extra chunks
+                local size = ("<I4"):unpack(handle.read(4))
+                handle.read(size)
+            end
+        until chunk == "data"
+
+        handle.read(4)
+        start = nil
+    end
+
     print("Playing " .. file)
 
-    local decoder = require "cc.audio.dfpwm".make_decoder()
+    local decoder = pcm and pcm_decoder or require "cc.audio.dfpwm".make_decoder()
```

Comment:
Seeing this makes me think that we should have a "cc.audio.pcm" with its own `make_decoder()`.

I have no idea how practical that might be though, I'm not an audio person.

Reply:
PCM decoding (at least unsigned 8-bit) is as simple as using `string.byte` and subtracting 128. Signed formats can all be accomplished with a loop of `string.unpack` and division. It's really very few lines of code - I don't really see the need for it (except maybe helping some newbies, but at that point just copy and paste it).

Also, no need to have encoder/decoder objects - PCM is stateless. DPCM/ADPCM ((adaptive) differential PCM) would, but those are not as efficient as DFPWM anyway.